python/dgl/dataloading/graphsaint.py

```diff
@@ -0,0 +1,128 @@
+"""GraphSAINT samplers."""
+from ..base import DGLError
+from ..random import choice
+from ..sampling import random_walk, pack_traces
+from .base import set_node_lazy_features, set_edge_lazy_features, Sampler
+
+try:
+    import torch
+except ImportError:
+    pass
+
+class SAINTSampler(Sampler):
+    """Random node/edge/walk sampler from
+    `GraphSAINT: Graph Sampling Based Inductive Learning Method
+    <https://arxiv.org/abs/1907.04931>`__
+
+    For each call, the sampler samples a node subset and then returns a node induced subgraph.
+    There are three options for sampling node subsets:
+
+    - For :attr:`'node'` sampler, the probability to sample a node is in proportion
+      to its out-degree.
+    - The :attr:`'edge'` sampler first samples an edge subset and then use the
+      end nodes of the edges.
+    - The :attr:`'walk'` sampler uses the nodes visited by random walks. It uniformly selects
+      a number of root nodes and then performs a fixed-length random walk from each root node.
+
+    Parameters
+    ----------
+    mode : str
+        The sampler to use, which can be :attr:`'node'`, :attr:`'edge'`, or :attr:`'walk'`.
+    budget : int or tuple[int]
+        Sampler configuration.
+
+        - For :attr:`'node'` sampler, budget specifies the number of nodes
+          in each sampled subgraph.
+        - For :attr:`'edge'` sampler, budget specifies the number of edges
+          to sample for inducing a subgraph.
+        - For :attr:`'walk'` sampler, budget is a tuple. budget[0] specifies
+          the number of root nodes to generate random walks. budget[1] specifies
+          the length of a random walk.
+
+    prefetch_ndata : list[str], optional
+        The node data to prefetch for the subgraph.
+
+        See :ref:`guide-minibatch-prefetching` for a detailed explanation of prefetching.
+    prefetch_edata : list[str], optional
+        The edge data to prefetch for the subgraph.
+
+        See :ref:`guide-minibatch-prefetching` for a detailed explanation of prefetching.
+    output_device : device, optional
+        The device of the output subgraphs.
+
+    Examples
+    --------
+
+    >>> import torch
+    >>> from dgl.dataloading import SAINTSampler, DataLoader
+    >>> num_iters = 1000
+    >>> sampler = SAINTSampler(mode='node', budget=6000)
+    >>> dataloader = DataLoader(g, torch.arange(num_iters), sampler, num_workers=4)
+    >>> for subg in dataloader:
+    ...     train_on(subg)
+    """
+    def __init__(self, mode, budget, prefetch_ndata=None,
+                 prefetch_edata=None, output_device='cpu'):
+        super().__init__()
```

Comment:
Since SAINTSampler only supports PyTorch, it's best to raise an error here if the backend is not PyTorch.

Reply:
It seems that all these samplers like `ClusterGCNSampler` and `ShaDowKHopSampler` are supposed to be used with `dgl.dataloading.DataLoader`, which inherits PyTorch's `DataLoader`.